lib/internal/modules/esm/get_format.js

```diff
@@ -92,6 +92,12 @@ let typelessPackageJsonFilesWarnedAbout;
 function getFileProtocolModuleFormat(url, context = { __proto__: null }, ignoreErrors) {
   const { source } = context;
   const ext = extname(url);
+  const deduceFormat = (fromSource, fromUrl) => {
+    const { existsSync, readFileSync } = require('fs');
+    const realSource = fromSource ?? existsSync(fromUrl) ? readFileSync(fromUrl).toString() : undefined;
```

Comment:
I'm no expert, but reading the file here might cause some performance issues

Reply:
of course it induces some performance penalty but since we need to solve the module format without having the source, the problem has no other solution. Returning some fictive `commonjs` is IMO wrong because that might or might not be confirmed by the `load` hook. IMO it's better to not answer a question than answering it with the wrong answer.
And the module detection already induces some performance penalty as it currently is implemented (it compiles the module to see if it contains module syntax). That's why we have the warning in place to advise users to define the packages/modules correctly so that this performance penalty price does not need to be paid.

Reply:
Thanks for the info! Over all, if the tests pass with and without `detect_module`, I am +1 (but I'm not a core collaborator, I'm just a triage helper)

Reply:
they don't pass with `true` and `false` with the same tests. Because without module detection you have different limitations and the tests have to account for those. With module detection on the situation is more flexible, more combinations are supported. The changes in the tests should explain this better than text can.  So those would be different tests (maybe in the same files but some logic to differentiate is needed because this is a change in behaviour as described in the PR description). 

The goal was to get them all pass with `detect_module` set to `true` and this is how they look like.

Reply:
Yes, I'm aware of that, but do the ones that are *expected* to pass pass?

Reply:
everything is green with `detect_module` `true` and some expectedly are now not green if that's set to `false`

Reply:
Woohoo!